### ISSUE TITLE
Universal build of llvm/lldb for macOS now requires 10.14 as deployment target

### ISSUE THREAD

**Issue by mstorsjo** [llvm/llvm-project#57017]
For redistribution of llvm and associated tools, I build it as universal binaries for macOS. I do this by passing `-DCMAKE_OSX_ARCHITECTURES=arm64;x86_64 -DCMAKE_OSX_DEPLOYMENT_TARGET=10.9` to CMake - so I simply do one build, where each object file is compiled for both architectures at once.

(A better universal build would be to do two full builds in parallel and then merge all the binaries with `lipo` - but that's much more effort. This kind of single-build universal binaries have some minor downsides, e.g. if building on x86_64, the arm64 version of the binaries still default to targeting `x86_64-apple-darwin`. But for a toolchain that primarily is meant for cross compiling targeting other OSes, this is a quite good compromise.)

Since switching to C++17, this build configuration fails with errors like these:
```console
lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp:641:7: error: aligned deallocation function of type 'void (void *, std::align_val_t) noexcept' is only available on macOS 10.14 or newer
class RegisterContextDarwin_arm64_Mach : public RegisterContextDarwin_arm64 {
      ^
```

The root cause for this seems to be that data types that require 16 byte alignment now triggers calls to aligned new/delete - which is only available since macOS 10.14. And this is only required for the arm64 part of the compilation - apparently, the regular new/delete is enough for 16 byte alignment on x86_64, but for arm64, it needs the newer aligned new/delete.

This is a somewhat silly situation, because in the universal binary, the x86_64 part of the binary, which doesn't need the modern aligned new/delete, is the part where it makes sense to target an older deployment target. For arm64, there's no need to target anything below 11.0 anyway (the first public release of arm64 macOS). If one would make the universal binaries by doing two separate compilations, it would be easy to set different deployment targets for them each - but if doing a single build for two architectures, it's no longer possible to target versions below 10.14.

This only seems to be an issue in LLDB - if only building e.g. llvm, lld, clang, it works fine to keep targeting older versions. Presumably, the issue could be also be avoided by not using types that requires larger alignments.

The core issue can be reproduced with a test snippet like this:
```c++
class MyClass {
public:
  virtual ~MyClass() {}
  __uint128_t var;
};
```
```console
$ clang++ -c aligned.cpp -c -std=c++17 -mmacosx-version-min=10.13 -arch arm64
aligned.cpp:4:11: error: aligned deallocation function of type 'void (void *, enum std::align_val_t) noexcept' is only available on macOS 10.14 or newer
  virtual ~MyClass() {}
          ^
aligned.cpp:4:11: note: if you supply your own aligned allocation functions, use -faligned-allocation to silence this diagnostic
1 error generated.
```
If either of the parameters is tweaked; changing to `-std=c++14`, setting `-mmacosx-version-min=10.14` or `-arch x86_64`, the compilation of the test snippet succeeds.

**Comment by mstorsjo** [comment]
It seems that this is only an issue when building with Xcode 12 (or contempory CommandLineTools). In Xcode 13, the default alignment for the regular, non-aligned, regular operator new/delete seems to be 16 on arm64 too, just like it was on x86_64 - so the code in LLDB no longer requires using the aligned operator new/delete in C++17, which makes it fine to keep the deployment target set to versions below 10.14.

**Comment by mstorsjo** [comment]
The upstream change that fixed this issue seems to have been this: aade0ec23b5986a9c478c4093d029a5db8a2c012

```
commit aade0ec23b5986a9c478c4093d029a5db8a2c012
Author: Akira Hatanaka <ahatanaka@apple.com>
Date:   Tue Feb 2 13:17:29 2021 -0800

    Fix the guaranteed alignment of memory returned by malloc/new on Darwin
    
    The guaranteed alignment is 16 bytes on Darwin.
    
    rdar://73431623
    
    Differential Revision: https://reviews.llvm.org/D95910
```

This later change (which doesn't seem to be in any shipped version of Xcode yet) also will affect these boundaries: 4722b5d1330a1104a0aa256804b9537c4d2c39d4

```
commit 4722b5d1330a1104a0aa256804b9537c4d2c39d4
Author: Louis Dionne <ldionne.2@gmail.com>
Date:   Tue Jul 5 18:53:33 2022 -0400

    [clang] Correct the macOS version that supports aligned allocation
    
    After checking the libc++abi.dylib shipped in macOS 10.13, I can confirm
    that it contains the align_val_t variants of operator new and operator
    delete. However, the libc++abi.dylib shipped on macOS 10.12 does not.
    
    Differential Revision: https://reviews.llvm.org/D129198
```

**Comment by mstorsjo** [comment]
Closing this issue, there's no action required here as this seems to be an issue primarily with Xcode 12.
